File: backend/app/models/plot.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import String, Boolean, DateTime, ForeignKey, Numeric, Text, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID

from app.core.database import Base


class Plot(Base):
    __tablename__ = "plots"

    id: Mapped[str] = mapped_column(UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4()))
    owner_id: Mapped[str] = mapped_column(ForeignKey("users.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    address: Mapped[str | None] = mapped_column(String(500))
    latitude: Mapped[float | None] = mapped_column(Numeric(10, 8))
    longitude: Mapped[float | None] = mapped_column(Numeric(11, 8))
    area_sqm: Mapped[float | None] = mapped_column(Numeric(10, 2))
    description: Mapped[str | None] = mapped_column(Text)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    # Связи
    user_roles: Mapped[list["UserPlotRole"]] = relationship(back_populates="plot")  # noqa
    zones: Mapped[list["Zone"]] = relationship(back_populates="plot", cascade="all, delete-orphan")  # noqa


class UserPlotRole(Base):
    __tablename__ = "user_plot_roles"

    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), primary_key=True)
    plot_id: Mapped[str] = mapped_column(ForeignKey("plots.id", ondelete="CASCADE"), primary_key=True)
    role: Mapped[str] = mapped_column(String(20), default="member")  # admin / member / viewer

    user: Mapped["User"] = relationship(back_populates="plot_roles")  # noqa
    plot: Mapped["Plot"] = relationship(back_populates="user_roles")  # noqa


class Zone(Base):
    __tablename__ = "zones"

    id: Mapped[str] = mapped_column(UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4()))
    plot_id: Mapped[str] = mapped_column(ForeignKey("plots.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    zone_type: Mapped[str] = mapped_column(String(50), default="garden")
    color: Mapped[str | None] = mapped_column(String(7))
    icon: Mapped[str | None] = mapped_column(String(50))
    description: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    plot: Mapped["Plot"] = relationship(back_populates="zones")  # noqa

File: mobile/lib/models/plot.dart
/// Участок (дача/сад/огород).
///
/// Зеркалит схему `PlotResponse` из backend (см. backend/app/schemas/plot.py).
/// Поля `latitude` / `longitude` в текущей версии API не возвращаются,
/// поэтому в модели их нет — добавим позже, когда подключим карту.
class Plot {
  final String id;
  final String name;
  final String? address;
  final double? areaSqm;
  final String? description;
  final bool isArchived;
  final DateTime createdAt;

  /// Роль текущего пользователя на этом участке: admin / member / viewer.
  /// Может быть null, если backend не вернул (на /plots возвращается всегда).
  final String? role;

  const Plot({
    required this.id,
    required this.name,
    this.address,
    this.areaSqm,
    this.description,
    required this.isArchived,
    required this.createdAt,
    this.role,
  });

  /// Парсит ответ от /api/v1/plots или /api/v1/plots/{id}.
  factory Plot.fromJson(Map<String, dynamic> json) {
    return Plot(
      id: json['id'] as String,
      name: json['name'] as String,
      address: json['address'] as String?,
      // area_sqm может прийти как int или double — приводим аккуратно.
      areaSqm: (json['area_sqm'] as num?)?.toDouble(),
      description: json['description'] as String?,
      isArchived: json['is_archived'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      role: json['role'] as String?,
    );
  }

  /// Удобно для UI: «У меня роль admin → могу редактировать».
  bool get isAdmin => role == 'admin';
}
